File: 98-validate-binary-search-tree/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return validate(root, nullptr, nullptr);
    }
    
private:
    bool validate(TreeNode* node, TreeNode* low, TreeNode* high) {
        // An empty tree/subtree is valid
        if (node == nullptr) {
            return true;
        }
        
        // The current node's value must be strictly between the low and high bounds
        if ((low != nullptr && node->val <= low->val) || 
            (high != nullptr && node->val >= high->val)) {
            return false;
        }
        
        // Recursively validate the left and right subtrees
        // For the left subtree, the current node becomes the upper bound (high)
        // For the right subtree, the current node becomes the lower bound (low)
        return validate(node->left, low, node) && 
               validate(node->right, node, high);
    }
};

File: 98-validate-binary-search-tree/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def validate(node, low=float('-inf'), high=float('inf')):
            # An empty tree/subtree is valid
            if not node:
                return True
            
            # The current node's value must be strictly between the low and high bounds
            if not (low < node.val < high):
                return False
            
            # Recursively validate the left and right subtrees
            # Left child inherits the current node's value as its upper bound
            # Right child inherits the current node's value as its lower bound
            return (validate(node.left, low, node.val) and 
                    validate(node.right, node.val, high))
        
        return validate(root)